qgrad/qgrad_qutip.py

```diff
@@ -377,3 +380,102 @@ def to_dm(state):
         )
 
     return out
+
+
+def make_rot(N, params, idx):
+    r"""Returns an :math:`N \times N` rotation matrix :math:`R_{ij}`,
+    where :math:`R_{ij}` is an :math:`N-`dimensional identity matrix
+    with the elements :math:`R_{ii}, R_{ij}, R_{ji}` and :math:`R_{jj}`
+    replaced as follows:
+
+    .. math::
+
+        \begin{pmatrix} R_{ii} & R{ij} \\ R_{ji} & R_{jj} 
+        \end{pmatrix} = \begin{pmatrix}
+            e^{i\phi_{ij}}cos(\theta_{ij}) & 
+            -e^{i\phi_{ij}sin(\theta_{ij})} \\
+            sin(\theta_{ij}) & cos(\theta_{ij})
+        \end{pmatrix}
+
+    Ref: Jing, Li, et al. "Tunable efficient unitary neural
+    networks (eunn) and their application to rnns."
+    International Conference on Machine Learning. 2017.
+    
+    Args:
+        N (int): dimension of the rotation matrix
+        params(:obj:`jnp.ndarray`): array of rotation parameters,
+                    :math:`\theta_{ij}` and :math:`\phi_{ij}` of
+                    shape (2, )
+        idx (tuple): indices (i, j) whose 4 permutations (as shown in
+                    the equation above) are to update the :math:`N \times N`
+                    identity to a rotation matrix by substituting `params`
+
+    Returns:
+        :obj:`jnp.ndarray`: :math:`N \times N` rotation matrix
+    """
+    i, j = idx
+    theta, phi = params
+    rotation = jnp.eye(N, dtype=jnp.complex64)
+    # updating the four entries
+    rotation = index_update(rotation, index[i, i], jnp.exp(1j * phi) * jnp.cos(theta))
+    rotation = index_update(rotation, index[i, j], -jnp.exp(1j * phi) * jnp.sin(theta))
+    rotation = index_update(rotation, index[j, i], jnp.sin(theta))
+    rotation = index_update(rotation, index[j, j], jnp.cos(theta))
+    return rotation
+
+
+def make_unitary(N, thetas, phis, omegas):
+    r"""Returns an :math:`N \times N` parameterized unitary 
+    matrix :math:`U(N)` using the following scheme
+        
+    .. math::
+        U(N) = D\prod_{i=2}^{N}\prod_{j=1}^{i-1}R^{'}_{ij}
+    
+    where :math:`D` is a diagonal matrix, whose elements are 
+    :math:`e^{i\omega{j}}` and :math:`R^{`}_{ij}` are rotation 
+    matrices (available via `make_rot`) where
+    :math:`R^{`}_{ij} = R(-\theta_{ij}, \phi_{ij})`
+        
+    .. note::
+        There are a total of :math:`\frac{N}(N-1)}{2}` 
+        :math:`\theta_{ij}` parameters :math:`\frac{N}(N-1)}{2}` 
+        :math:`\phi{ij}` parameters, and :math:`N omega_{ij}`
+        parameters 
+        
+    Args:
+        N (int): Dimension of the unitary matrix
+        thetas (:obj:`jnp.ndarray`): theta angles for rotations
```

Comment:
This is well written. Maybe in order to be consistent with the other major operator in this library `Displace()`, we should go for a `Unitary()` class which has the same signature as displace. You construct a N dim unitary as `u = Unitary(N)` first, and then apply it such that `u(thetas, phis, omegas)` uses the `__call__` method to give the unitary.